File: module_examples/vp/test_mpp_encoder.cpp
#include "module/vi/module_memReader.hpp"
#include "module/vp/module_mppenc.hpp"

// #define TEST_PARALLEL_ENCODING

#ifdef TEST_PARALLEL_ENCODING
#include "module/vp/module_rga.hpp"
#endif

void status_change_callback(void* ctx, ModuleStatus status)
{
    ff_info("Module state has changed(%d)\n", status);
}

void output_callback(void* ctx, shared_ptr<MediaBuffer> buffer)
{
    FILE* fp = (FILE*)ctx;
    fwrite(buffer->getActiveData(), buffer->getActiveSize(), 1, fp);

    static uint64_t frame_count;
    static auto start_time = std::chrono::high_resolution_clock::now();
    if (++frame_count % 100 == 0) {
        auto diff = std::chrono::duration_cast<std::chrono::milliseconds>(
                        std::chrono::high_resolution_clock::now() - start_time)
                        .count();
        ff_info("encode %ld frames time %ld ms fps %ld \n",
                frame_count, diff, frame_count * 1000 / diff);
    }
}

void fill_nv12_image(shared_ptr<VideoBuffer> buffer, int buffer_index)
{
    int x, y, i;
    auto image_param = buffer->getImagePara();
    int width = image_param.width;
    int height = image_param.height;
    int h_stride = image_param.hstride;
    int v_stride = image_param.vstride;

    uint8_t* data_y = (uint8_t*)buffer->getActiveData();
    uint8_t* data_c = data_y + h_stride * v_stride;
    i = buffer_index;

    for (y = 0; y < height; y++)
        for (x = 0; x < width; x++)
            data_y[y * h_stride + x] = x + y + i * 3;

    for (y = 0; y < height / 2; y++) {
        for (x = 0; x < width; x++) {
            data_c[y * h_stride + x++] = 128 + y + i * 2;
            data_c[y * h_stride + x] = 128 + x + i * 5;
        }
    }
}

int main(int argc, char** argv)
{
    int ret;
    std::shared_ptr<ModuleMedia> last_mod;

    if (argc < 2) {
        ff_error("\nUsage: %s output.h265\n", argv[0]);
        return -1;
    };

    auto fp = fopen(argv[1], "w+");
    if (fp == nullptr) {
        ff_error("Failed to open %s: %s\n", argv[1], strerror(errno));
        return -1;
    }

    /* Prepare a dummy input image. */
    ImagePara input_para = ImagePara(1920, 1080, 1920, 1080,
                                     V4L2_PIX_FMT_NV12);
    auto input_buffer = make_shared<VideoBuffer>(
        VideoBuffer::DRM_BUFFER_NONCACHEABLE);
    input_buffer->allocBuffer(input_para);
    if (input_buffer->getActiveData() == nullptr) {
        ff_error("Failed to alloc buffer\n");
        fclose(fp);
        return -1;
    }
    fill_nv12_image(input_buffer, 0);

    /* Create a memory reader module. */
    auto r_mem = make_shared<ModuleMemReader>(input_buffer->getImagePara());
    r_mem->setStatusChangeCallback(nullptr, status_change_callback);
    ret = r_mem->init();
    if (ret < 0) {
        ff_error("Failed to init memory reader, %d\n", ret);
        fclose(fp);
        return ret;
    }
    last_mod = r_mem;

#ifdef TEST_PARALLEL_ENCODING
    /* Copy the data to more buffer queues to improve encoding parallelism. */
    auto rga = make_shared<ModuleRga>(last_mod->getOutputImagePara(),
                                      RGA_ROTATE_NONE);
    rga->setProductor(last_mod);
    /* Set the buffer queue length to 5. */
    rga->setBufferCount(5);
    ret = rga->init();
    if (ret < 0) {
        ff_error("Failed to init rga converter, %d\n", ret);
        fclose(fp);
        return ret;
    }
    last_mod = rga;
#endif

    /* Create a mpp encoder module. */
    auto v_enc = make_shared<ModuleMppEnc>(EncodeType::ENCODE_TYPE_H265);
    v_enc->setProductor(last_mod);
    ret = v_enc->init();
    if (ret < 0) {
        ff_error("Failed to init mpp encoder, %d\n", ret);
        fclose(fp);
        return ret;
    }

    /* Print the video stream information */
    auto v_extra = v_enc->getExtraBuffer();
    if (v_extra) {
        ff_info("Video: Extra codec %d , data %p, bytes %ld\n",
                v_extra->getMediaCodec(), v_extra->getActiveData(),
                v_extra->getActiveSize());

        auto param = v_extra->getImagePara();
        ff_info("foramt %s, width %d, height %d\n",
                v4l2GetFmtName(param.v4l2Fmt), param.width,
                param.height);
    }

    /* Set the output callback function for the encoder */
    v_enc->setOutputDataCallback(fp, output_callback);

    ff_info("\n========================================================\n\n");

    /* The start module will start all consumer modules under the module and
        all consumer modules under the consumer module. */
    r_mem->start();
    int frame_count = 0;
    while (frame_count++ < 2000) {
        /* There is no need to draw images in real time when testing the
            encoding rate. */
#ifndef TEST_PARALLEL_ENCODING
        fill_nv12_image(input_buffer, frame_count);
#endif

        ret = r_mem->setInputBuffer(input_buffer);
        if (ret != 0) {
            ff_error("Failed to set the input buffer\n");
            break;
        }

        ret = r_mem->waitProcess(2000);
        if (ret != 0) {
            ff_warn("Wait timeout\n");
            break;
        }
    }

    r_mem->setProcessStatus(ModuleMemReader::PROCESS_STATUS_EXIT);

    /* The stop module will stop all consumer modules under the module and all
        consumer modules under the consumer module. */
    r_mem->stop();
    fclose(fp);
    ff_info("Exit.\n");
    return 0;
}

File: module_examples/vp/test_mpp_encoder.py
#!/usr/bin/env python3

import sys
import time
import ff_pymedia as ff

test_parallel_encoding = 1
input_frame_counter = 0
output_frame_counter = 0
init_time = time.time()
loop_time = time.time()

def output_callback(file, buffer):
    global output_frame_counter, loop_time

    output_frame_counter += 1

    if output_frame_counter % 100 == 0:
        duration = time.time() - loop_time
        fps = round(100 / duration)
        loop_time = time.time()
        print(f"encode 100 frames time {duration} fps: {fps}")

    # get memoryview object
    data = buffer.getActiveData()
    file.write(data)
    #print("Data size ", data.nbytes)

def status_change_callback(obj, status):
    print(f"{obj} module state has changed: ", status)


def fill_nv12_image(buffer, buffer_index):
    image_param = buffer.getImagePara()
    width = image_param.width
    height = image_param.height
    h_stride = image_param.hstride
    v_stride = image_param.vstride

    # get memoryview object
    data = buffer.getActiveData()
    i = buffer_index

    # Y plane fill
    for y in range(height):
        y_offset = y * h_stride
        for x in range(width):
            data[y_offset + x] = (x + y + i * 3) & 0xFF

    # UV plane fill
    uv_start = h_stride * v_stride
    for y in range(height // 2):
        uv_offset = uv_start + y * h_stride
        for x in range(0, width, 2):
            data[uv_offset + x] = (128 + y + i * 2) & 0xFF      # U
            data[uv_offset + x + 1] = (128 + x + i * 5) & 0xFF  # V


def main():
    global input_frame_counter, output_frame_counter, init_time, loop_time, test_parallel_encoding
    if len(sys.argv) > 1:
        file_path = sys.argv[1]
    else:
        file_path = input("Please enter the output file path: ")

    file = open(file_path, "wb")

    # Prepare a dummy input image.
    image_param = ff.ImagePara(1920, 1080, 1920, 1080, ff.v4l2GetFmtByName("NV12"))
    input_buffer = ff.VideoBuffer(ff.DRM_BUFFER_NONCACHEABLE)
    input_buffer.allocBuffer(image_param)
    if input_buffer.getSize() == 0:
        print("Failed to alloc buffer")
        return -1
    fill_nv12_image(input_buffer, 0)

    # Create a memory reader module.
    mem_reader = ff.ModuleMemReader(input_buffer.getImagePara())
    mem_reader.setStatusChangeCallback("Memory reader", status_change_callback)
    ret = mem_reader.init()
    if ret < 0:
        print("Failed to init memory reader, ", ret)
        return ret

    last_mod = mem_reader
    if test_parallel_encoding != 0:
        # Copy the data to more buffer queues to improve encoding parallelism.
        rga_converter = ff.ModuleRga(last_mod.getOutputImagePara(), ff.RGA_ROTATE_NONE)
        rga_converter.setProductor(last_mod)
        # Set the buffer queue length to 5.
        rga_converter.setBufferCount(5)
        ret = rga_converter.init()
        if ret < 0:
            print("Failed to init rga converter, ", ret)
            return ret

        last_mod = rga_converter


    # Create a mpp encoder module.
    v_encoder = ff.ModuleMppEnc(ff.ENCODE_TYPE_H265)
    v_encoder.setProductor(last_mod)
    # Set the callback function for the encoder.
    v_encoder.setOutputDataCallback(file, output_callback)
    ret = v_encoder.init()
    if ret < 0:
        print("Failed to init mpp encoder, ", ret)


    print("\n============================START=============================\n")
    mem_reader.start()
    init_time = time.time()
    loop_time = init_time
    while input_frame_counter < 2000:
        ret = mem_reader.setInputBuffer(input_buffer)
        if ret != 0:
            print("Failed to set the input buffer, ", ret)
            break

        ret = mem_reader.waitProcess(2000)
        if ret != 0:
            print("Wait timeout")
            break

        input_frame_counter += 1
        # There is no need to draw images in real time when testing the
        # encoding rate.
        #if test_parallel_encoding == 0:
        #    fill_nv12_image(input_buffer, input_frame_counter)

    mem_reader.setProcessStatus(ff.PROCESS_STATUS_EXIT)
    mem_reader.stop()
    file.close()
    print("\n============================STOP==============================\n")
    print(f"Input frames {input_frame_counter}, output frames {output_frame_counter}")
    print("Average fps: ", input_frame_counter / (time.time() - init_time))

if __name__ == "__main__":
    main()
